### ISSUE TITLE
Popen is not Sync on windows

### ISSUE THREAD

**Issue by palaviv** [hniksic/rust-subprocess#42]
```rust
help: within `subprocess::popen::Popen`, the trait `std::marker::Sync` is not implemented for `*mut bitflags::core::ffi::c_void`
   = note: required because it appears within the type `subprocess::win32::Handle`
   = note: required because it appears within the type `subprocess::popen::os::ExtChildState`
   = note: required because it appears within the type `subprocess::popen::ChildState`
   = note: required because it appears within the type `subprocess::popen::Popen`
```
From looking at the code in win32.rs there is:
 ```rust
unsafe impl Send for Handle {}
```
My guess is that this should be solved by adding:
```rust
unsafe impl Sync for Handle {}
```

**Comment by hniksic** [comment]
Thanks for the report. Do you have example code to reproduce this issue?

**Comment by palaviv** [comment]
```rust
struct Popen {
    process: subprocess::Popen
}

impl Sync for Popen {}
```
Running this on Windows should reproduce this.

**Comment by hniksic** [comment]
Thanks for the snippet, but I was aiming for an example that shows what you wanted to achieve by sharing `Popen` between threads. `Sync` doesn't seem very useful for `Popen` since `Sync` gives you the guarantee that `&self` methods are thread-safe, and most of `Popen`'s public methods take `&mut self`. In fact, sharing `Popen` between threads guarantees that there will be no mutable reference to `Popen`, so you won't be able to do anything useful with the value.

Since `Popen` uses no interior mutability, there should be no adverse effects of it implementing `Sync`, but I'd still like to understand your usage. Without an example, it seems that you're not really gaining anything from the change.

**Comment by hniksic** [comment]
Hi @palaviv, any news on this? Do you have an example of why you need `Popen` to be `Sync`, given that most of its methods take `&mut self`?

**Comment by palaviv** [comment]
Hi @hniksic sorry for the delayed response. We use `rust-suborocess` to implement the subprocess library in RustPython. Here you can see our [Popen](https://github.com/RustPython/RustPython/blob/802913b8499bbb6a08512f58ba3af71493eee957/vm/src/stdlib/subprocess.rs#L18) object. We wrap the `Popen` in RwLock in order to get `&mut self`.
The current [documentation](https://docs.rs/subprocess/0.2.4/subprocess/struct.Popen.html#synthetic-implementations) indicate that `impl Sync for Popen`. I think that using `Popen` from multiple threads should be possible as this is possible in python:
```python
import threading
import subprocess

r = subprocess.Popen(["sleep" , "5"])

def thread():
    print("status:", r.poll())
    print("done:", r.wait())


t1 = threading.Thread(target=thread)
t2 = threading.Thread(target=thread)
t1.start()
t2.start()
t1.join()
t2.join()
```

**Comment by hniksic** [comment]
> We wrap the `Popen` in `RwLock` in order to get `&mut self`.

That sounds correct. My point is that, even if `Popen` were `Sync`, you'd still need a lock to access it from multiple threads. This is because `Sync` only means that `&self` methods can be called concurrently, and most of `Popen`'s methods take `&mut self`.

So I can make the change to make it more consistent and to match the documentation, but as far as I can tell, it won't make the least bit of difference for your use case.

**Comment by hniksic** [comment]
Fixed in subprocess 0.2.6, with the caveat mentioned in the previous comment.
